include/seqan3/alignment/matrix/alignment_trace_matrix.hpp

```diff
@@ -98,75 +99,148 @@ struct alignment_coordinate
+    typename gapped_query_alphabet_t = gapped<std::decay_t<decltype(std::declval<query_t>()[0])>>>
 //!\cond
     requires matrix_concept<remove_cvref_t<trace_matrix_t>> &&
              std::Same<typename remove_cvref_t<trace_matrix_t>::entry_type, trace_directions>
 //!\endcond
-inline std::pair<std::list<gapped_database_t>, std::list<gapped_query_t>>
-alignment_trace(database_t && database, query_t && query, trace_matrix_t && matrix, alignment_coordinate const end_coordinate)
+inline std::pair<std::vector<gapped_database_alphabet_t>, std::vector<gapped_query_alphabet_t>>
+alignment_trace(database_t && database,
+                query_t && query,
+                trace_matrix_t && matrix,
+                alignment_coordinate const end_coordinate)
 {
+    using signed_size_t = std::make_signed_t<size_t>;
+
     constexpr auto N = trace_directions::none;
     constexpr auto D = trace_directions::diagonal;
     constexpr auto L = trace_directions::left;
     constexpr auto U = trace_directions::up;
-    size_t col = end_coordinate.seq1_pos + 1;
-    size_t row = end_coordinate.seq2_pos + 1;
+    signed_size_t col = end_coordinate.seq1_pos + 1;
+    signed_size_t row = end_coordinate.seq2_pos + 1;
 
-    assert(col <= database.size());
     assert(row <= query.size());
+    assert(col <= database.size());
+    assert(row < matrix.rows());
+    assert(col < matrix.cols());
 
-    std::list<gapped_database_t> gapped_database{};
-    std::list<gapped_query_t> gapped_query{};
+    std::deque<gapped_database_alphabet_t> gapped_database{};
+    std::deque<gapped_query_alphabet_t> gapped_query{};
 
     if (matrix.at(0, 0) != N)
-        throw std::runtime_error{"End trace must be NONE"};
+        throw std::logic_error{"End trace must be NONE"};
 
-    while(row >= 0 && col >= 0)
+    while(true)
     {
         trace_directions dir = matrix.at(row, col);
         if ((dir & L) == L)
         {
-            col = std::max<int>(col - 1, 0);
+            col = std::max<signed_size_t>(col - 1, 0);
             gapped_database.push_front(database[col]);
             gapped_query.push_front(gap::GAP);
         }
         else if ((dir & U) == U)
         {
-            row = std::max<int>(row - 1, 0);
+            row = std::max<signed_size_t>(row - 1, 0);
             gapped_database.push_front(gap::GAP);
             gapped_query.push_front(query[row]);
         }
         else if ((dir & D) == D)
         {
-            row = std::max<int>(row - 1, 0);
-            col = std::max<int>(col - 1, 0);
+            row = std::max<signed_size_t>(row - 1, 0);
+            col = std::max<signed_size_t>(col - 1, 0);
             gapped_database.push_front(database[col]);
             gapped_query.push_front(query[row]);
         }
         else
         {
             if (row == 0 || col == 0)
                 break;
-            throw std::runtime_error{"Trace not found"};
+            throw std::logic_error{"Trace not found"};
```

Comment:
This is not a very expressive error message IMPOV. Maybe `Unkown seqan3::trace_direction in an inner cell of the trace matrix.`